src/libfsm/determinise.c

```diff
@@ -1261,59 +1251,1014 @@ analyze_closures__analyze(struct analyze_closures_env *env)
+{
+	assert(0 == (state_id & RESULT_BIT));
+	return pair_cache_save(env, RPBT_SINGLE_STATE, state_id, state_id, result_id);
+}
+
+static int
+analysis_cache_save_pair(struct analyze_closures_env *env,
+	fsm_state_t a, fsm_state_t b, fsm_state_t result_id)
+{
+	assert(a != b);
+	return pair_cache_save(env, RPBT_PAIR, a | RESULT_BIT, b | RESULT_BIT, result_id);
+}
+
+static size_t pair_cache_hit, pair_cache_miss;
+
+static int
+combine_pair(struct analyze_closures_env *env, fsm_state_t pa, fsm_state_t pb,
+	fsm_state_t *result_id)
+{
+#if LOG_GROUPING
+	fprintf(stderr, "%s: pa %d, pb %d\n", __func__, pa, pb);
+#endif
+
+	/* If the IDs are for single states, check if they've already been
+	 * analyzed. If not, do that and cache it. Either way, switch to the
+	 * ID for their cached analysis result. */
+	if (0 == (pa & RESULT_BIT)) {
+		fsm_state_t cached_result_id;
+		if (!analysis_cache_check_single_state(env, pa, &cached_result_id)) {
+			if (!cache_single_state_analysis(env, pa, &cached_result_id)) {
+				return 0;
+			}
+		}
+		pa = cached_result_id | RESULT_BIT;
+	}
+
+	if (0 == (pb & RESULT_BIT)) {
+		fsm_state_t cached_result_id;
+		if (!analysis_cache_check_single_state(env, pb, &cached_result_id)) {
+			if (!cache_single_state_analysis(env, pb, &cached_result_id)) {
+				return 0;
+			}
+		}
+		pb = cached_result_id | RESULT_BIT;
+	}
+
+	if (analysis_cache_check_pair(env, pa, pb, result_id)) {
+#if LOG_GROUPING
+		fprintf(stderr, "%s: cache hit for { %d%s %d%s }: %d\n",
+		    __func__, ID_WITH_SUFFIX(pa), ID_WITH_SUFFIX(pb), *result_id);
+#endif
+		pair_cache_hit++;
+		return 1;
+	} else {
+#if LOG_GROUPING
+		fprintf(stderr, "%s: cache miss for { %d%s %d%s }\n",
+		    __func__, ID_WITH_SUFFIX(pa), ID_WITH_SUFFIX(pb));
+#endif
+		pair_cache_miss++;
+	}
+
+	assert((pa & RESULT_BIT) && (pb & RESULT_BIT));
+	pa &=~ RESULT_BIT;
+	pb &=~ RESULT_BIT;
+
+	if (!combine_result_pair_and_commit(env, pa, pb, result_id)) {
+		return 0;
+	}
+
+#if LOG_GROUPING
+	fprintf(stderr, "%s: combined and committed as result_id %d\n",
+	    __func__, *result_id);
+#endif
+
+	if (!analysis_cache_save_pair(env, pa, pb, *result_id)) {
+		return 0;
+	}
+
+	return 1;
+}
+
+#if EXPENSIVE_CHECKS
+static void
+assert_entry_labels_are_mutually_exclusive(const struct analysis_result *r)
+{
+	uint64_t seen[4] = { 0 };
+	for (size_t e_i = 0; e_i < r->count; e_i++) {
+		const struct result_entry *e = &r->entries[e_i];
+		const uint64_t *labels = e->labels;
+		for (size_t i = 0; i < 4; i++) {
+			assert((seen[i] & labels[i]) == 0);
+			seen[i] |= labels[i];
+		}
+	}
+}
+#endif
+
+static int
+cmp_fsm_state_t(const void *pa, const void *pb)
+{
+	const fsm_state_t a = *(const fsm_state_t *)pa;
+	const fsm_state_t b = *(const fsm_state_t *)pb;
+	return a < b ? -1 : a > b ? 1 : 0;
+}
+
+static void
+sort_and_dedup_dst_buf(fsm_state_t *buf, size_t *used)
+{
+	const size_t orig_used = *used;
+	qsort(buf, orig_used, sizeof(buf[0]), cmp_fsm_state_t);
```

Comment:
Unfortunately this adds a `qsort` on the per-DFA-state path. It could be worth replacing that with a less general sort, since we know the input must be a set of valid state IDs `0 <= id < fsm->statecount` and any duplicates can be rejected. That's better off happening in a different PR, though.